app/models/ems_refresh/save_inventory_helper.rb

```diff
@@ -113,12 +113,28 @@ def save_child_inventory(obj, hashes, child_keys, *args)
 
   def store_ids_for_new_records(records, hashes, keys)
     keys = Array(keys)
-    hashes.each do |h|
-      r = records.detect { |r| keys.all? { |k| r.send(k) == r.class.type_for_attribute(k.to_s).cast(h[k]) } }
-      h[:id] = r.id
+    # Lets first index the hashes based on keys, so we can do O(1) lookups
+    record_index = {}
+    records.each do |record|
```

Comment:
@cben so the records are already in the memory, so we should not fetch them again if possible. But then we should not even keep all the records in memory. But saying that, we should not be doing this post indexing where we go through all of it again. :-)

The answer is, I rewrote it all for the graph refresh. :-) For the old refresh, I am just trying to eliminate O(n^2) and similar issues, that breaks the refresh after reaching some amount of data. But to actually optimize memory and speed, it needs a total rewrite, like the graph refresh does. :-)